### PR TITLE
Bootstrap from C#. Number formats

### PR DESCRIPTION
**Pull request by iskiselev**
Per discussion in #850.
Added new C# project - JSIL.mscorlib.
This project is translated on JSIL compilation and goes into Libraries folder. It is attached as bootstrap through JSIL.js.
Currently I've added (Numeric).ToString(format, ...) implementations through mono System.NumberFormatter.

Also this PR contains several StringBuilder method implementations, fix for #855.

To support Bootstrap from C# there was introduced several meta attributes:
JSRepaceAssemblyDeclaration - replaces JSIL.DeclareAssembly with passed string 
JSOverrideAssemblyReference - overrides $asm... for passed type assembly with string and enclose all translation in self-execution function
JSImportType - include type from other assembly in translation as it was defined here. Should be use with JSProxy.
JSSuppressTypeDeclaration (#507) - TypeDeclarationsToSuppress moved here.

@kg, please wait for tests to pass. I've not yet check all of them - but we can already start review of this change.

### CODE REVIEW

File: JSIL.Libraries/Includes/Bootstrap/Resources/Main.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ if (!$jsilcore)
 
 $jsilcore.getCurrentUICultureImpl = function () {
   var language;
-  var svc = JSIL.Host.getService("window", false);
+  var svc = JSIL.Host.getService("window", true);
```
Comment (kg) [inline JSIL.Libraries/Includes/Bootstrap/Resources/Main.js]: Why?

Comment (iskiselev) [inline JSIL.Libraries/Includes/Bootstrap/Resources/Main.js]: We need initialize current culture even in non-browser environment.
Really next line even tries to do it:
```
  if (svc) {
    language = svc.getNavigatorLanguage() || "en-US";
  } else {
    language = "en-US";
  }
```

But in non-browser environment it will not return null, but fail.
Now it will return null.

Comment (kg) [inline JSIL.Libraries/Includes/Bootstrap/Resources/Main.js]: Ah, I probably got the second argument's boolean nature backwards... bools are awful.

**Comment by iskiselev** [comment]
Currently JSImport works incorrectly on Linux.
Some methods of imported classes become UntranslatableFunctions. Right now will look on it.
I've now worked on Linux a lot before. @kg, what could you advice to debug JSIL on Linux (except of improving log output)?

**Comment by kg** [comment]
I don't really know, I find Linux development a nightmare. None of the tools are good. That's why some of the 'doesn't work on Mono' tests are still broken.

**Comment by iskiselev** [comment]
@kg, how could function go into FunctionCache?
On linux, when I'm try to translate imported functions, looks like some of them are already in cache with null records.
I start suppose that either JSIL key comparer works incorrectly - or Concurrent Dictionary returns incorrect records.

**Comment by iskiselev** [comment]
It was my fault. Functions went to cache through GetSecondPass.
Last that I still don't understand - why it wasn't happen on windows (both mono and .net)?

**Comment by kg** [comment]
Mono's ConcurrentDictionary is probably broken. They've had trivial bugs in their BCL containers before, and they take forever to get fixed (even when I provide a fix). It's a nightmare.

**Comment by iskiselev** [comment]
Formatting test on travis still failing, but I've just tested it on my own Linux box - and it works as expected.
For now I'll mark this test (IntegerToStringFormatted) as FailsOnMono.

We need really upload somewhere all artifacts (at least Libraries folder and all tests *.out files) to be able test problems that happens on Travis - but it is another task.

**Comment by iskiselev** [comment]
So, now I think this PR is stable enough and could be merged.
Probably we need write some additional checks for JSImport edge cases, but it may be better to do later - when we'll have some more experience with translated Bootstrap.

**Comment by iskiselev** [comment]
Ops. Looks like I missed non-integer types. Let's see test results.

**Comment by iskiselev** [comment]
Get artifacts from travis and looked why test fails there.
Looks like NumberFormatter.Format, that contains switch with goto translated incorrectly if was compiled under mono.

**Comment by iskiselev** [comment]
@kg, could you review/discuss this? I'm working now on new implementation for #773, and it is done on top of it. It will be not so easy to make that PR fully independent, but I don't want make this PR too complicated.

File: JSIL.mscorlib/JSIL/System/NumberFormatter.Data.cs [hunk 2]
```diff
@@ -0,0 +1,916 @@
+﻿//
+// mono/mono/metadata/number-formatter.h
```
Comment (kg) [inline JSIL.mscorlib/JSIL/System/NumberFormatter.Data.cs]: Please explicitly call out what license that was under in addition to the filename. Doesn't need the whole license prologue

File: JSIL/JavascriptFormatter.cs [hunk 3]
```diff
@@ -1164,10 +1167,30 @@ select GetNameOfType(ga as dynamic)).ToArray())
                 Space();
                 Identifier(PrivateToken.IDString);
                 WriteRaw(" = ");
-                WriteRaw("JSIL.DeclareAssembly");
-                LPar();
-                Value(Assembly.FullName);
-                RPar();
+                if (AssmeblyDeclarationReplacement == null)
```
Comment (kg) [inline JSIL/JavascriptFormatter.cs]: Can you fix that typo?

File: Proxies/BCL/System.cs [hunk 4]
```diff
@@ -28,4 +31,31 @@ public class System_Security_Cryptography_X509Certificates_X509Certificate2
...
+    [JSProxy(typeof(DateTimeFormatInfo), JSProxyMemberPolicy.ReplaceDeclared, JSProxyAttributePolicy.ReplaceDeclared, JSProxyInterfacePolicy.ReplaceNone, false)]
+    public class System_Globalization_DateTimeFormatInfo
+    {
+        [JSReplaceConstructor]
+        static System_Globalization_DateTimeFormatInfo()
+        {
+        }
+    }
+
+    /*[JSProxy(typeof(ExecutionContext), JSProxyMemberPolicy.ReplaceNone, JSProxyAttributePolicy.ReplaceDeclared, JSProxyInterfacePolicy.ReplaceNone, false)]
```
Comment (kg) [inline Proxies/BCL/System.cs]: Why is this commented out?

**Comment by kg** [comment]
Took a second look at this and I don't see anything wrong other than the two nits I commented on. Thanks.

File: JSIL.mscorlib/JSIL/System/NumberFormatter.Data.cs (hunk 2 above)
Comment (iskiselev) [inline JSIL.mscorlib/JSIL/System/NumberFormatter.Data.cs]: Hmm, good catch. Looks like this file, as it located in mono/mono folder is licensed only with LGPL or commercial license.
It is really bad, so we should check if it was previously located under anther directory pr ask them for using this file with another license :(

File: Proxies/BCL/System.cs (hunk 4 above)
Comment (iskiselev) [inline Proxies/BCL/System.cs]: Looks like I've added it by mistake. Removed.

**Comment by iskiselev** [comment]
PR fix applied. Also added double signature fix and .Parse(string, IFormatProvider) signature as it is used by NumberFormatter.
